### PR TITLE
Decode cookies safely

### PR DESCRIPTION
**Pull request by mungodewar**
👋  Hi folks, 

Currently using appium to run real device tests, however, running into issues with incorrectly encoded cookie values causing difficulty. On inspection of the code I found that there was no guard against a malformed cookie value. 

I've raised this PR to start a conversation if this is a change that you folks would consider? 

Many thanks!

### CODE REVIEW

File: lib/commands/cookies.js [hunk 1]
```diff
@@ -13,11 +13,18 @@ commands.getCookies = async function getCookies () {
...
-      value: decodeURI(cookie.value),
-    });
+  let decodedCookies = [];
+  cookies.cookies.forEach(function mapCookie (cookie) {
+    try {
+      decodedCookies.push(Object.assign({}, cookie, {
+        value: decodeURI(cookie.value),
+      }));
+    } catch (error) {
+      log.debug(`Cookie ${cookie.name} was not decoded successfully. Cookie value: ${cookie.value}`);
```
Comment (mykola-mokhnach) [inline lib/commands/cookies.js:23]: I would rather perform value decoding separately and catch exceptions there. e.g:

```
const decodedCookieValues = cookies.cookies.map((cookie) => {
  try {
    return decodeURI(cookie.value);
  } catch (e) {
    log.debug(`Cookie ${cookie.name} was not decoded successfully. Cookie value: ${cookie.value}`);
    log.error(e);
    return undefined;
  }
});
return _.zip(cookies.cookies, decodedCookieValues)
  .map(([cookie, value]) => Object.assign({}, cookie, _.isUndefined(value) ? {} : {value}));
```

Comment (mungodewar) [reply lib/commands/cookies.js:21]: Sounds good. I've separated as you indicated.

File: lib/commands/cookies.js [hunk 2]
```diff
@@ -11,20 +12,22 @@ commands.getCookies = async function getCookies () {
 
   // get the cookies from the remote debugger, or an empty object
   const cookies = await this.remote.getCookies() || {cookies: []};
-  // the value is URI encoded, so decode it
-  // but keep all the rest of the info intact
-  let decodedCookies = [];
-  cookies.cookies.forEach(function mapCookie (cookie) {
+
+  // the value is URI encoded, so decode it safely
+  const decodedCookieValues = cookies.cookies.map(function mapCookieValue (cookie) {
```
Comment (mykola-mokhnach) [inline lib/commands/cookies.js:17]: is there any particular reason to not use anonymous function here?

File: lib/commands/cookies.js [hunk 3]
```diff
@@ -11,20 +12,22 @@ commands.getCookies = async function getCookies () {
...
-      log.error(error);
+      log.warn(error);
+      return undefined;
     }
   });
-  return decodedCookies;
+
+  // zip cookes with decoded value, removing undefined cookie values
+  return _.zip(cookies.cookies, decodedCookieValues)
+  .filter(([, value]) => !_.isUndefined(value))
```
Comment (mykola-mokhnach) [inline lib/commands/cookies.js:29]: please add two spaces before starting dots, so it's easier to parse

File: lib/commands/cookies.js (hunk 2 above)
Comment (mungodewar) [reply lib/commands/cookies.js:17]: I don't think so. I was only replicating the style of the original. Happy to change.

Comment (mykola-mokhnach) [reply lib/commands/cookies.js:17]: I personally find it cleaner to use lambas unless we really need to use syntactical `this`

Comment (mungodewar) [reply lib/commands/cookies.js]: Yeah, totally agree. 👍

**Comment by mungodewar** [comment]
> @mungodewar Could you please sign the CLA?

Yes. Let me look into that 👀

**Review by mykola-mokhnach — APPROVED** [review]

**Comment by mykola-mokhnach** [comment]
Any updates @mungodewar 
I'm happy to merge the PR as soon as CLA is signed

**Comment by mungodewar** [comment]
> Any updates @mungodewar
> I'm happy to merge the PR right after the CLA is signed

Sorry for taking a while, should be all sorted now. Thanks!

Any idea when this might be cut for release?

**Comment by KazuCocoa** [comment]
We can create a new version of xcuitest-driver for you. The version will be available only `appium@beta` or Appium 2.0 (and install the xcuitest version) for now.

**Pull request merged**